### PR TITLE
Correctly discard only inactive streams.

### PR DESCRIPTION
**Pull request by Lukasa**
Motivation:

Long-lived connections that have many inactive streams would cause
crashes. Crashes are, in my professional opinion, less than ideal
when they occur under expected usage cases.

Modifications:

- Fixed an inverted boolean check that would discard only *active*
  streams.
- Added tests.

Result:

Fewer crashes, better software.

### CODE REVIEW

File: Sources/NIOHTTP2/HTTP2Parser.swift [hunk 1]
```diff
@@ -156,16 +156,27 @@ public final class HTTP2Parser: ChannelInboundHandler, ChannelOutboundHandler {
...
         self.mode = mode
         self.initialSettings = initialSettings
+        self.maxCachedClosedStreams = maxCachedClosedStreams
     }
 
     public func channelActive(ctx: ChannelHandlerContext) {
         self.session = NGHTTP2Session(mode: self.mode,
                                       allocator: ctx.channel.allocator,
-                                      maxCachedStreamIDs: 1024,  // TODO(cory): Make configurable
+                                      maxCachedStreamIDs: self.maxCachedClosedStreams,  // TODO(cory): Make configurable
```
Comment (weissi) [inline Sources/NIOHTTP2/HTTP2Parser.swift]: well, it's now configurable so the comment can go right? Or did you mean `ChannelOption` kind of configurable?

File: Tests/NIOHTTP2Tests/SimpleClientServerTests.swift [hunk 2]
```diff
@@ -1230,4 +1230,90 @@ class SimpleClientServerTests: XCTestCase {
         XCTAssertNoThrow(try self.clientChannel.finish())
         XCTAssertNoThrow(try self.serverChannel.finish())
     }
+
+    func testManyConcurrentInactiveStreams() throws {
+        let maxCachedClosedStreams = 1024
+
+        // Begin by getting the connection up.
+        try self.basicHTTP2Connection(maxCachedClosedStreams: 1024)
```
Comment (weissi) [inline Tests/NIOHTTP2Tests/SimpleClientServerTests.swift]: use the `maxCachedClosedStreams` variable here?

File: Tests/NIOHTTP2Tests/SimpleClientServerTests.swift [hunk 3]
```diff
@@ -1230,4 +1230,90 @@ class SimpleClientServerTests: XCTestCase {
...
+        XCTAssertNoThrow(try self.clientChannel.finish())
+        XCTAssertNoThrow(try self.serverChannel.finish())
+    }
+
+    func testDontRemoveActiveStreams() throws {
+        // This is a bit of a regression test, not a generally useful one.
+        let maxCachedClosedStreams = 1024
+
+        // Begin by getting the connection up.
+        try self.basicHTTP2Connection(maxCachedClosedStreams: 1024)
```
Comment (weissi) [inline Tests/NIOHTTP2Tests/SimpleClientServerTests.swift]: and here

File: Sources/NIOHTTP2/NGHTTP2Session.swift [hunk 4]
```diff
@@ -285,7 +285,7 @@ fileprivate struct StreamManager {
     // Discard old and unnecessary streams.
     private mutating func purgeOldStreams() {
         while self.streamMap.count >= maxSize {
-            let lowestStreamID = self.streamMap.filter { $0.value.active }.keys.sorted().first { $0 != 0 && $0 != Int32.max }!
+            let lowestStreamID = self.streamMap.filter { !$0.value.active }.keys.sorted().first { $0 != 0 && $0 != Int32.max }!
```
[the remark this answers is not in the document]
Comment (MrMage) [reply Sources/NIOHTTP2/NGHTTP2Session.swift]: I assume the filtered array (and, by extension, `self.streamMap`) will now never be empty?

File: Tests/NIOHTTP2Tests/SimpleClientServerTests.swift (hunk 2 above)
Comment (Lukasa) [reply Tests/NIOHTTP2Tests/SimpleClientServerTests.swift]: Fixed.

File: Tests/NIOHTTP2Tests/SimpleClientServerTests.swift (hunk 3 above)
Comment (Lukasa) [reply Tests/NIOHTTP2Tests/SimpleClientServerTests.swift]: Fixed.

File: Tests/NIOHTTP2Tests/SimpleClientServerTests.swift [hunk 5]
```diff
@@ -1230,4 +1230,90 @@ class SimpleClientServerTests: XCTestCase {
         XCTAssertNoThrow(try self.clientChannel.finish())
         XCTAssertNoThrow(try self.serverChannel.finish())
     }
+
+    func testManyConcurrentInactiveStreams() throws {
+        let maxCachedClosedStreams = 1024
+
+        // Begin by getting the connection up.
+        try self.basicHTTP2Connection(maxCachedClosedStreams: maxCachedClosedStreams)
```
Comment (MrMage) [inline Tests/NIOHTTP2Tests/SimpleClientServerTests.swift]: I was about to mention the following (first part is already fixed): As you are declaring `maxCachedClosedStreams` above, should you replace `1024` with `maxCachedClosedStreams` here? Also, would it make sense to pick a lower value than 1024 for faster test runtime?

File: Tests/NIOHTTP2Tests/SimpleClientServerTests.swift [hunk 6]
```diff
@@ -1230,4 +1230,90 @@ class SimpleClientServerTests: XCTestCase {
...
+
+            let serverStreamID = try self.assertFramesRoundTrip(frames: [reqFrame, reqBodyFrame], sender: self.clientChannel, receiver: self.serverChannel).first!.streamID
+
+            // Let's send a quick response back.
+            var respFrame = HTTP2Frame(streamID: serverStreamID, payload: .headers(responseHeaders))
+            respFrame.endStream = true
+            try self.assertFramesRoundTrip(frames: [respFrame], sender: self.serverChannel, receiver: self.clientChannel)
+        }
+
+        // Ok, now we're going to open *two* streams. In the old, broken code, the opening of the second
```
Comment (MrMage) [inline Tests/NIOHTTP2Tests/SimpleClientServerTests.swift]: Mention this PR to more easily reference what the broken code was?

File: Sources/NIOHTTP2/NGHTTP2Session.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (Lukasa) [reply Sources/NIOHTTP2/NGHTTP2Session.swift]: Not quite, but I wanted to separate the other issue out into another patch.

The TL;DR there is that if the sum of both sides SETTINGS_MAX_CONCURRENT_STREAMS is higher than `maxSize`, and there are actually more than `maxSides` valid open active streams, then we'll explode again. That can be fixed by changing the loop conditional above to move the `filter` out of the loop so that we only care if there are too many *closed* streams.

This will all get a bit better when we land #10 and the follow-on framing logic such that I can start work on replacing the nghttp2 state machine. At that stage we'll have a much more complete insight into the connection state and won't have to use this somewhat insane purging construction.

**Comment by Lukasa** [comment]
No need to apologise @MrMage, more code review is always valuable. I've addressed both your comments.

**Comment by Lukasa** [comment]
@weissi Want to re-review?

**Comment by bmhatfield** [comment]
(I'm just a bystander, but) Bump: @weissi this PR fixes an issue discovered in https://github.com/grpc/grpc-swift/pull/281 - just adding a friendly bump for a re-review here :-)
